File: listas.py
from nodos import Nodo


class ListaSimple:
    def __init__(self):
        self.cabeza = None

    # Recorre los nodos
    def recorrer(self):
        actual = self.cabeza
        while actual:
            yield actual
            actual = actual.siguiente

    # Inserta al final
    def insertar(self, valor):
        nuevo = Nodo(valor)
        if self.cabeza is None:
            self.cabeza = nuevo
        else:
            actual = self.cabeza
            while actual.siguiente:
                actual = actual.siguiente
            actual.siguiente = nuevo
        print("Elemento insertado.")

    # Elimina el primer valor encontrado
    def eliminar(self, valor):
        actual = self.cabeza
        anterior = None
        while actual and actual.valor != valor:
            anterior = actual
            actual = actual.siguiente
        if actual is None:
            print("Elemento no encontrado.")
            return
        print(f"Elemento a eliminar: {actual.valor}")
        if anterior is None:
            self.cabeza = actual.siguiente
        else:
            anterior.siguiente = actual.siguiente
        print("Elemento eliminado.")

    # Busca usando un for
    def buscar(self, valor):
        for posicion, actual in enumerate(self.recorrer()):
            if actual.valor == valor:
                print(f"Encontrado en posicion {posicion}.")
                return
        print("No encontrado.")

    # Muestra usando un for
    def mostrar(self):
        datos = []
        for actual in self.recorrer():
            datos.append(actual.valor)
        print(f"Lista: {datos}")


class ListaDoble:
    def __init__(self):
        self.cabeza = None
        self.cola = None

    # Recorre hacia adelante
    def recorrer_adelante(self):
        actual = self.cabeza
        while actual:
            yield actual
            actual = actual.siguiente

    # Recorre hacia atras
    def recorrer_atras(self):
        actual = self.cola
        while actual:
            yield actual
            actual = actual.anterior

    # Inserta al final
    def insertar(self, valor):
        nuevo = Nodo(valor)
        if self.cabeza is None:
            self.cabeza = nuevo
            self.cola = nuevo
        else:
            self.cola.siguiente = nuevo
            nuevo.anterior = self.cola
            self.cola = nuevo
        print("Elemento insertado.")

    # Elimina el primer valor encontrado
    def eliminar(self, valor):
        actual = None
        for nodo in self.recorrer_adelante():
            if nodo.valor == valor:
                actual = nodo
                break
        if actual is None:
            print("Elemento no encontrado.")
            return
        print(f"Elemento a eliminar: {actual.valor}")
        if actual.anterior:
            actual.anterior.siguiente = actual.siguiente
        else:
            self.cabeza = actual.siguiente
        if actual.siguiente:
            actual.siguiente.anterior = actual.anterior
        else:
            self.cola = actual.anterior
        print("Elemento eliminado.")

    # Muestra hacia adelante con for
    def mostrar_adelante(self):
        datos = []
        for actual in self.recorrer_adelante():
            datos.append(actual.valor)
        print(f"Adelante: {datos}")

    # Muestra hacia atras con for
    def mostrar_atras(self):
        datos = []
        for actual in self.recorrer_atras():
            datos.append(actual.valor)
        print(f"Atras: {datos}")

File: version_c/listas.c
#include <stdio.h>
#include <stdlib.h>
#include "listas.h"

// Prepara la lista
void inicializar_lista_simple(ListaSimple *lista) {
    lista->cabeza = NULL;
}

// Inserta al final
void insertar_simple(ListaSimple *lista, int valor) {
    NodoSimple *nuevo = (NodoSimple *) malloc(sizeof(NodoSimple));
    nuevo->valor = valor;
    nuevo->siguiente = NULL;

    if (lista->cabeza == NULL) {
        lista->cabeza = nuevo;
    } else {
        NodoSimple *actual = lista->cabeza;
        while (actual->siguiente != NULL) {
            actual = actual->siguiente;
        }
        actual->siguiente = nuevo;
    }
    printf("Elemento insertado\n");
}

// Elimina un valor
void eliminar_simple(ListaSimple *lista, int valor) {
    NodoSimple *actual = lista->cabeza;
    NodoSimple *anterior = NULL;

    while (actual != NULL && actual->valor != valor) {
        anterior = actual;
        actual = actual->siguiente;
    }

    if (actual == NULL) {
        printf("Elemento no encontrado\n");
        return;
    }

    printf("Elemento a eliminar: %d\n", actual->valor);
    if (anterior == NULL) {
        lista->cabeza = actual->siguiente;
    } else {
        anterior->siguiente = actual->siguiente;
    }
    free(actual);
    printf("Elemento eliminado\n");
}

// Busca un valor
void buscar_simple(ListaSimple *lista, int valor) {
    NodoSimple *actual = lista->cabeza;
    int posicion = 0;
    while (actual != NULL) {
        if (actual->valor == valor) {
            printf("Encontrado en posicion %d\n", posicion);
            return;
        }
        actual = actual->siguiente;
        posicion++;
    }
    printf("No encontrado\n");
}

// Muestra la lista
void mostrar_simple(ListaSimple *lista) {
    NodoSimple *actual = lista->cabeza;
    printf("Lista: [");
    while (actual != NULL) {
        printf("%d", actual->valor);
        if (actual->siguiente != NULL) {
            printf(" ");
        }
        actual = actual->siguiente;
    }
    printf("]\n");
}

// Libera memoria
void liberar_lista_simple(ListaSimple *lista) {
    while (lista->cabeza != NULL) {
        NodoSimple *temporal = lista->cabeza;
        lista->cabeza = lista->cabeza->siguiente;
        free(temporal);
    }
}

// Prepara la lista
void inicializar_lista_doble(ListaDoble *lista) {
    lista->cabeza = NULL;
    lista->cola = NULL;
}

// Inserta al final
void insertar_doble(ListaDoble *lista, int valor) {
    NodoDoble *nuevo = (NodoDoble *) malloc(sizeof(NodoDoble));
    nuevo->valor = valor;
    nuevo->siguiente = NULL;
    nuevo->anterior = NULL;

    if (lista->cabeza == NULL) {
        lista->cabeza = nuevo;
        lista->cola = nuevo;
    } else {
        lista->cola->siguiente = nuevo;
        nuevo->anterior = lista->cola;
        lista->cola = nuevo;
    }
    printf("Elemento insertado\n");
}

// Elimina un valor
void eliminar_doble(ListaDoble *lista, int valor) {
    NodoDoble *actual = lista->cabeza;

    while (actual != NULL && actual->valor != valor) {
        actual = actual->siguiente;
    }

    if (actual == NULL) {
        printf("Elemento no encontrado\n");
        return;
    }

    printf("Elemento a eliminar: %d\n", actual->valor);
    if (actual->anterior != NULL) {
        actual->anterior->siguiente = actual->siguiente;
    } else {
        lista->cabeza = actual->siguiente;
    }

    if (actual->siguiente != NULL) {
        actual->siguiente->anterior = actual->anterior;
    } else {
        lista->cola = actual->anterior;
    }

    free(actual);
    printf("Elemento eliminado\n");
}

// Muestra hacia adelante
void mostrar_doble_adelante(ListaDoble *lista) {
    NodoDoble *actual = lista->cabeza;
    printf("Adelante: [");
    while (actual != NULL) {
        printf("%d", actual->valor);
        if (actual->siguiente != NULL) {
            printf(" ");
        }
        actual = actual->siguiente;
    }
    printf("]\n");
}

// Muestra hacia atras
void mostrar_doble_atras(ListaDoble *lista) {
    NodoDoble *actual = lista->cola;
    printf("Atras: [");
    while (actual != NULL) {
        printf("%d", actual->valor);
        if (actual->anterior != NULL) {
            printf(" ");
        }
        actual = actual->anterior;
    }
    printf("]\n");
}

// Libera memoria
void liberar_lista_doble(ListaDoble *lista) {
    while (lista->cabeza != NULL) {
        NodoDoble *temporal = lista->cabeza;
        lista->cabeza = lista->cabeza->siguiente;
        free(temporal);
    }
    lista->cola = NULL;
}
